x/ccv/child/keeper/distribution.go

```diff
@@ -0,0 +1,114 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	transfertypes "github.com/cosmos/ibc-go/v3/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v3/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v3/modules/core/04-channel/types"
+
+	"github.com/cosmos/interchain-security/x/ccv/child/types"
+	ccv "github.com/cosmos/interchain-security/x/ccv/types"
+)
+
+// Simple model, send tokens to the fee pool of the provider validator set
+// reference: cosmos/ibc-go/v3/modules/apps/transfer/keeper/msg_server.go
+func (k Keeper) DistributeToProviderValidatorSet(ctx sdk.Context) error {
+
+	ltbh, err := k.GetLastTransmissionBlockHeight(ctx)
+	if err != nil {
+		return err
+	}
+	bpdt := k.GetBlocksPerDistributionTransmission(ctx)
+	curHeight := ctx.BlockHeight()
+
+	if (curHeight - ltbh.Height) < bpdt {
+		// not enough blocks have passed for  a transmission to occur
+		return nil
+	}
+
+	// work around to reuse the IBC token transfer logic
+	consumerFeePoolAddr := k.accountKeeper.GetModuleAccount(ctx, k.feeCollectorName).GetAddress()
+	fpTokens := k.bankKeeper.GetAllBalances(ctx, consumerFeePoolAddr)
+
+	// split the fee pool, send the consumer's fraction to the consumer redistribution address
+	fracStr := k.GetConsumerRedistributeFrac(ctx)
+	frac, err := sdk.NewDecFromStr(fracStr)
+	if err != nil {
+		return err
+	}
+	decFPTokens := sdk.NewDecCoinsFromCoins(fpTokens...)
+	consRedistrTokens, _ := decFPTokens.MulDec(frac).TruncateDecimal()
+	err = k.bankKeeper.SendCoinsFromModuleToModule(ctx, k.feeCollectorName, types.ConsumerRedistributeName, consRedistrTokens)
+	if err != nil {
+		return err
+	}
+
+	// Send the remainder to the Provider fee pool over ibc
+	remainingTokens := fpTokens.Sub(consRedistrTokens)
+	ch := k.GetDistributionTransmissionChannel(ctx)
+	providerAddr := k.GetProviderFeePoolAddrStr(ctx)
+	timeoutHeight := clienttypes.Height{0, 0}
+	timeoutTimestamp := uint64(ccv.GetTimeoutTimestamp(ctx.BlockTime()).UnixNano())
```

Comment:
We should use a different timeout for token transfer than for CCV packets. For CCV packets we use large timeouts to ensure that the CCV channel will not timeout and close (since it's an ordered channel). In the case a token transfer times out, the transfer channel is not closed and the funds are returned to `consumerFeePoolAddr`. 

These funds will be sent again with the next transfer. However, these funds shouldn't be multiplied again with `ConsumerRedistributeFrac`.  This means that the `consumerFeePool` should collect only the funds that need to be transferred to the provider.

Reply:
What is the benefit of this channel timing out? For the regular transfer use case, when a human is trying to transfer, it is so that they can do something else with their money if the transfer doesn't work. For this, it seems like it would be ok for packets to stay valid as long as possible.

Reply:
Yeah, but what if nobody relays that transfer ever? The transfer channel is unordered, which means that a transfer can be skipped. A timeout fixes that.

Reply:
Agreed. We can use a smaller timeout. And since the money is refunded to the same address, we effectively retry the transfer in the next epoch (along with the new accumulated funds).

By the way, this is an additional reason to do the splitting of funds elsewhere, and send full account balance to provider

Reply:
First note: I've updated the timeout to a hardcoded calculation set to 1 week (instead of 4 weeks as it was before). @mpoke if you think this should be a param or some other value, just lmn. 

Secondly: Great catch for the rejected transfer situation. The solution is really easy, which is to send the tokens to a buffer address on the consumer chain side. If the transfer times out, it just gets sent back to the buffer where it _will not_ have a second redistribution split applied to it when it is sent out with the next distribution event. Personally I don't see this as a maintenance item. It's literally 9 lines of code right now + a few comments. I think it's valuable to potential consumer chains to have this boilerplate split written out, I imagine most chains will want to have this split functionality whereby they can choose. 

Re: @AdityaSripal (replying to your comment in this thread) 

> Why are we doing this logic here? It seems simpler to me that the child chain application decides how much of the fee goes into their community pool, and how much goes into the ccv account. And then the full ccv account gets sent to the provider. It also reduces the code inside ccv

So this location _is_ the first point in which fees/minted-tokens are first processes and where the fundamental split is made between consumer tokens and provider tokens. The design decision here is that all further use/distribution of funds on the consumer chains should be made _from_ the consumer's redistribute address, whether that be providing the tokens to a consumer chain community pool, developer multi-sig or whatever, I don't think any of those splits should happen before the "major" split between what's staying on the chain and what's going to the provider chain. In this way, this logic is actually consistent with your comment "that the child chain application decides how much of the fee goes into their community pool" We're not writing logic here about how a consumer chain should use its funds, just providing the infrastructure to make a split at this fundamental level. I also don't see it as a maintenance item given the simplicity of the solution. If we don't provide this basic split functionality in one way or another, the ccv distribution feels a bit "half-baked" to me.

Reply:
I agree with Rigel here. The nice thing about the way he's done things is that it requires no changes to Cosmos-SDK, just app.go. I think having more code in CCV is a good thing because it is easier for us to work with here and we are the one building the feature. It is also good for the user because the functionality is closely related to IS, so it keeps all the params and stuff related to it in one place.

I would recommend that he do the "buffer account" to avoid the double splitting and keep the current splitting logic in CCV.

Reply:
I agree. The buffer account will solve the issue. Also, having the split happing in the CCV module seems to me cleaner since it's part of the CCV protocol. 

Regarding the transfer timeout, I'm not sure whether 1 week is still too long. By [default](https://github.com/cosmos/interchain-security/blob/ab153fceccac12b7957d73e5401785f7f758b3e1/x/ccv/child/types/params.go#L37), the consumer is sending rewards every 2h. Keeping this interval small reduces the inaccuracy of the simple distribution (i.e., the rewards are distributed to the validators that validate at the time the transfer is received). For the same reason, the timeout should also be smaller. Having it too small though will lead to a lot of timeout packets that need to be relay. So maybe 1 day would be decent. What do you say @AdityaSripal? 

In my view, the only reason timeouts will be common is if the amount of tokens transferred in each packet is very small, i.e., the value of the packet is too low to even bother relaying it. In that case, relayers would ignore these packets and wait for more rewards to accumulate. However, timing out packets also requires relaying TimeoutMsgs, so I'm not sure if that would make sense financially.